File: FBDeviceControl/Management/FBDeviceDebugServer.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBDeviceDebugServer.h"

#import <FBControlCore/FBControlCore.h>

#import "FBAMDServiceConnection.h"

@interface FBDeviceDebugServer_TwistedPairFiles : NSObject

@property (nonatomic, assign, readonly) int socket;
@property (nonatomic, strong, readonly) FBAMDServiceConnection *connection;
@property (nonatomic, strong, readonly) id<FBControlCoreLogger> logger;
@property (nonatomic, strong, readonly) dispatch_queue_t socketToConnectionQueue;
@property (nonatomic, strong, readonly) dispatch_queue_t connectionToSocketQueue;

@end

@implementation FBDeviceDebugServer_TwistedPairFiles

- (instancetype)initWithSocket:(int)socket connection:(FBAMDServiceConnection *)connection logger:(id<FBControlCoreLogger>)logger
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _socket = socket;
  _connection = connection;
  _logger = logger;
  _socketToConnectionQueue = dispatch_queue_create("com.facebook.fbdevicecontrol.debugserver.socket_to_connection", DISPATCH_QUEUE_SERIAL);
  _connectionToSocketQueue = dispatch_queue_create("com.facebook.fbdevicecontrol.debugserver.connection_to_socket", DISPATCH_QUEUE_SERIAL);

  return self;
}

static size_t const ConnectionReadSizeLimit = 1024;

- (FBFuture<NSNull *> *)startWithError:(NSError **)error
{
  if (@available(macOS 10.15, *)) {
    id<FBControlCoreLogger> logger = self.logger;
    int socket = self.socket;
    NSFileHandle *socketReadHandle = [[NSFileHandle alloc] initWithFileDescriptor:socket closeOnDealloc:NO];
    NSFileHandle *socketWriteHandle = [[NSFileHandle alloc] initWithFileDescriptor:socket closeOnDealloc:NO];
    FBAMDServiceConnection *connection = self.connection;
    FBMutableFuture<NSNull *> *socketReadCompleted = FBMutableFuture.future;
    FBMutableFuture<NSNull *> *connectionReadCompleted = FBMutableFuture.future;
    dispatch_async(self.socketToConnectionQueue, ^{
      while (socketReadCompleted.state == FBFutureStateRunning && connectionReadCompleted.state == FBFutureStateRunning) {
        NSError *innerError = nil;
        NSData *data = [socketReadHandle availableData];
        if (data.length == 0) {
          [logger log:@"Socket read reached end of file"];
          break;
        }
        if (![connection send:data error:&innerError]) {
          [logger logFormat:@"Sending data to remote debugserver failed: %@", innerError];
          break;
        }
      }
      [logger logFormat:@"Exiting socket %d read loop", socket];
      [socketReadCompleted resolveWithResult:NSNull.null];
    });
    dispatch_async(self.connectionToSocketQueue, ^{
      while (socketReadCompleted.state == FBFutureStateRunning && connectionReadCompleted.state == FBFutureStateRunning) {
        NSError *innerError = nil;
        NSData *data = [connection receiveUpTo:ConnectionReadSizeLimit error:&innerError];
        if (data.length == 0) {
          [logger logFormat:@"debugserver read ended: %@", innerError];
          break;
        }
        if (![socketWriteHandle writeData:data error:&innerError]) {
          [logger logFormat:@"Socket write failed: %@", innerError];
          break;
        }
      }
      [logger logFormat:@"Exiting connection %@ read loop", connection];
      [connectionReadCompleted resolveWithResult:NSNull.null];
    });
    return [[FBFuture
      futureWithFutures:@[
        socketReadCompleted,
        connectionReadCompleted,
      ]]
      onQueue:self.connectionToSocketQueue notifyOfCompletion:^(id _) {
        [logger logFormat:@"Closing socket file descriptor %d", socket];
        close(socket);
      }];
  }
  return nil;
}

@end

@interface FBDeviceDebugServer () <FBSocketServerDelegate>

@property (nonatomic, strong, readonly) FBAMDServiceConnection *serviceConnection;
@property (nonatomic, strong, readonly) FBSocketServer *tcpServer;
@property (nonatomic, strong, readonly) id<FBControlCoreLogger> logger;

@property (nonatomic, strong, readwrite) FBMutableFuture<NSNull *> *teardown;
@property (nonatomic, strong, nullable, readwrite) FBDeviceDebugServer_TwistedPairFiles *twistedPair;

@end

@implementation FBDeviceDebugServer

@synthesize queue = _queue;
@synthesize lldbBootstrapCommands = _lldbBootstrapCommands;

#pragma mark Initializers

+ (FBFuture<FBDeviceDebugServer *> *)debugServerForServiceConnection:(FBFutureContext<FBAMDServiceConnection *> *)service port:(in_port_t)port lldbBootstrapCommands:(NSArray<NSString *> *)lldbBootstrapCommands queue:(dispatch_queue_t)queue logger:(id<FBControlCoreLogger>)logger
{
  return [[service
    onQueue:queue push:^(FBAMDServiceConnection *serviceConnection) {
      FBDeviceDebugServer *server = [[FBDeviceDebugServer alloc] initWithServiceConnection:serviceConnection port:port lldbBootstrapCommands:lldbBootstrapCommands queue:queue logger:logger];
      return [server startListening];
    }]
    onQueue:queue enter:^(FBDeviceDebugServer *server, FBMutableFuture<NSNull *> *teardown) {
      server.teardown = teardown;
      return server;
    }];
}

- (instancetype)initWithServiceConnection:(FBAMDServiceConnection *)serviceConnection port:(in_port_t)port lldbBootstrapCommands:(NSArray<NSString *> *)lldbBootstrapCommands queue:(dispatch_queue_t)queue logger:(id<FBControlCoreLogger>)logger
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _serviceConnection = serviceConnection;
  _tcpServer = [FBSocketServer socketServerOnPort:port delegate:self];
  _lldbBootstrapCommands = lldbBootstrapCommands;
  _queue = queue;
  _logger = logger;

  return self;
}

#pragma mark FBSocketReaderDelegate

- (void)socketServer:(FBSocketServer *)server clientConnected:(struct in6_addr)address fileDescriptor:(int)fileDescriptor
{
  if (self.twistedPair) {
    [self.logger log:@"Rejecting connection, we have an existing pair"];
    NSData *data = [@"$NEUnspecified#00" dataUsingEncoding:NSASCIIStringEncoding];
    write(fileDescriptor, data.bytes, data.length);
    close(fileDescriptor);
    return;
  }
  [self.logger log:@"Client connected, connecting all file handles"];
  FBDeviceDebugServer_TwistedPairFiles *twistedPair = [[FBDeviceDebugServer_TwistedPairFiles alloc] initWithSocket:fileDescriptor connection:self.serviceConnection logger:self.logger];
  NSError *error = nil;
  FBFuture<NSNull *> *completed = [twistedPair startWithError:&error];
  if (!completed) {
    [self.logger logFormat:@"Failed to start connection %@", error];
    return;
  }
  [completed onQueue:self.queue notifyOfCompletion:^(id _) {
    [self.logger log:@"Client Disconnected"];
    self.twistedPair = nil;
  }];
  [self.teardown resolveFromFuture:completed];
  self.twistedPair = twistedPair;
}

#pragma mark FBiOSTargetOperation

- (FBFuture<NSNull *> *)completed
{
  return self.teardown;
}

#pragma mark Private Methods

- (FBFutureContext<FBDeviceDebugServer *> *)startListening
{
  return [[self.tcpServer
    startListeningContext]
    onQueue:self.queue pend:^(NSNull *_) {
      [self.logger logFormat:@"TCP Server now running, boostrap commands for lldb are %@", [self.lldbBootstrapCommands componentsJoinedByString:@"\n"]];
      return [FBFuture futureWithResult:self];
    }];
}


@end

File: FBDeviceControl/Management/FBDeviceDebugServer.swift
// (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.

@preconcurrency import FBControlCore
import Foundation

private let connectionReadSizeLimit: size_t = 1024

private class FBDeviceDebugServer_TwistedPairFiles: NSObject {
  let socket: Int32
  let connection: FBAMDServiceConnection
  let logger: any FBControlCoreLogger
  let socketToConnectionQueue: DispatchQueue
  let connectionToSocketQueue: DispatchQueue

  init(
    socket: Int32,
    connection: FBAMDServiceConnection,
    logger: any FBControlCoreLogger
  ) {
    self.socket = socket
    self.connection = connection
    self.logger = logger
    self.socketToConnectionQueue = DispatchQueue(label: "com.facebook.fbdevicecontrol.debugserver.socket_to_connection")
    self.connectionToSocketQueue = DispatchQueue(label: "com.facebook.fbdevicecontrol.debugserver.connection_to_socket")
    super.init()
  }

  func start() -> FBFuture<NSNull>? {
    guard #available(macOS 10.15, *) else {
      return nil
    }

    let logger = self.logger
    let socket = self.socket
    let socketReadHandle = FileHandle(fileDescriptor: socket)
    nonisolated(unsafe) let connection = self.connection
    let socketReadCompleted = FBMutableFuture<NSNull>()
    let connectionReadCompleted = FBMutableFuture<NSNull>()

    socketToConnectionQueue.async {
      while socketReadCompleted.state == .running && connectionReadCompleted.state == .running {
        let data = socketReadHandle.availableData
        if data.isEmpty {
          logger.log("Socket read reached end of file")
          break
        }
        do {
          try connection.send(data as Data)
        } catch {
          logger.log("Sending data to remote debugserver failed: \(error)")
          break
        }
      }
      logger.log("Exiting socket \(socket) read loop")
      socketReadCompleted.resolve(withResult: NSNull())
    }

    connectionToSocketQueue.async {
      while socketReadCompleted.state == .running && connectionReadCompleted.state == .running {
        do {
          let data = try connection.receiveUp(to: connectionReadSizeLimit)
          if data.count == 0 {
            logger.log("debugserver read ended")
            break
          }
          data.withUnsafeBytes { bufferPointer in
            guard let baseAddress = bufferPointer.baseAddress else { return }
            var totalWritten = 0
            while totalWritten < data.count {
              let written = Darwin.write(socket, baseAddress.advanced(by: totalWritten), data.count - totalWritten)
              if written <= 0 {
                logger.log("Socket write failed")
                return
              }
              totalWritten += written
            }
          }
        } catch {
          logger.log("debugserver read ended: \(error)")
          break
        }
      }
      logger.log("Exiting connection \(connection) read loop")
      connectionReadCompleted.resolve(withResult: NSNull())
    }

    let socketFuture = unsafeBitCast(socketReadCompleted, to: FBFuture<NSNull>.self)
    let connectionFuture = unsafeBitCast(connectionReadCompleted, to: FBFuture<NSNull>.self)
    let raceFuture = FBFuture<NSNull>(race: [socketFuture, connectionFuture])
      .onQueue(
        connectionToSocketQueue,
        doOnResolved: { _ in
          logger.log("Closing socket file descriptor \(socket)")
          close(socket)
        }
      )
      .mapReplace(NSNull())
    return unsafeBitCast(raceFuture, to: FBFuture<NSNull>.self)
  }
}

@objc(FBDeviceDebugServer)
public class FBDeviceDebugServer: NSObject, FBSocketServerDelegate, FBDebugServer {
  private let serviceConnection: FBAMDServiceConnection
  private lazy var tcpServer: FBSocketServer = FBSocketServer(onPort: self.port, delegate: self)
  private let port: in_port_t
  private let logger: any FBControlCoreLogger
  private var teardown: FBMutableFuture<NSNull>?
  private var twistedPair: FBDeviceDebugServer_TwistedPairFiles?

  // MARK: - FBDebugServer

  @objc public let lldbBootstrapCommands: [String]

  // MARK: - FBSocketServerDelegate

  @objc public let queue: DispatchQueue

  // MARK: - Initializers

  /// Factory method: creates and starts a debug server from a future-wrapped service connection.
  @objc public static func debugServer(
    forServiceConnection service: FBFutureContext<FBAMDServiceConnection>,
    port: in_port_t,
    lldbBootstrapCommands: [String],
    queue: DispatchQueue,
    logger: (any FBControlCoreLogger)?
  ) -> FBFuture<AnyObject> {
    return service.onQueue(
      queue,
      push: { (serviceConnection: AnyObject) -> FBFutureContext<AnyObject> in
        let connection = serviceConnection as! FBAMDServiceConnection
        let server = FBDeviceDebugServer(
          serviceConnection: connection,
          port: port,
          lldbBootstrapCommands: lldbBootstrapCommands,
          queue: queue,
          logger: logger
        )
        return server.startListening() as! FBFutureContext<AnyObject>
      }
    ).onQueue(
      queue,
      enter: { (result: AnyObject, teardownFuture: FBMutableFuture<NSNull>) -> AnyObject in
        let server = result as! FBDeviceDebugServer
        server.teardown = teardownFuture
        return server
      })
  }

  init(
    serviceConnection: FBAMDServiceConnection,
    port: in_port_t,
    lldbBootstrapCommands: [String],
    queue: DispatchQueue,
    logger: (any FBControlCoreLogger)?
  ) {
    self.serviceConnection = serviceConnection
    self.port = port
    self.lldbBootstrapCommands = lldbBootstrapCommands
    self.queue = queue
    self.logger = logger ?? FBControlCoreGlobalConfiguration.defaultLogger
    super.init()
  }

  // MARK: - FBSocketServerDelegate

  @objc public func socketServer(
    _ server: FBSocketServer,
    clientConnected address: in6_addr,
    fileDescriptor: Int32
  ) {
    if twistedPair != nil {
      logger.log("Rejecting connection, we have an existing pair")
      if let data = "$NEUnspecified#00".data(using: .ascii) {
        data.withUnsafeBytes { bufferPointer in
          guard let baseAddress = bufferPointer.baseAddress else { return }
          _ = Darwin.write(fileDescriptor, baseAddress, data.count)
        }
      }
      close(fileDescriptor)
      return
    }
    logger.log("Client connected, connecting all file handles")
    let pair = FBDeviceDebugServer_TwistedPairFiles(
      socket: fileDescriptor,
      connection: serviceConnection,
      logger: logger
    )
    guard let completed = pair.start() else {
      logger.log("Failed to start connection")
      return
    }
    completed.onQueue(
      queue,
      doOnResolved: { [weak self] _ in
        self?.logger.log("Client Disconnected")
        self?.twistedPair = nil
      })
    teardown?.resolve(from: unsafeBitCast(completed, to: FBFuture<AnyObject>.self))
    self.twistedPair = pair
  }

  // MARK: - FBiOSTargetOperation

  @objc public var completed: FBFuture<NSNull> {
    if let teardown = teardown {
      return unsafeBitCast(teardown, to: FBFuture<NSNull>.self)
    }
    return unsafeBitCast(FBMutableFuture<NSNull>(), to: FBFuture<NSNull>.self)
  }

  // MARK: - Private Methods

  private func startListening() -> FBFutureContext<FBDeviceDebugServer> {
    return tcpServer.startListeningContext()
      .onQueue(
        queue,
        pend: { [self] (_: AnyObject) -> FBFuture<AnyObject> in
          self.logger.log("TCP Server now running, bootstrap commands for lldb are \(self.lldbBootstrapCommands.joined(separator: "\n"))")
          return FBFuture<AnyObject>(result: self)
        }) as! FBFutureContext<FBDeviceDebugServer>
  }
}
